core/src/main/java/com/graphhopper/routing/util/spatialrules/AbstractSpatialRule.java

```diff
@@ -52,6 +53,11 @@ public RoadAccess getAccess(RoadClass roadClass, TransportationMode transport, R
         return currentRoadAccess;
     }
     
+    @Override
+    public Toll getToll(RoadClass roadClass, TransportationMode transport, Toll currentToll) {
+        return currentToll;
+    }
+    
```

Comment:
Hmmh it would be really nice if we do not need to introduce a new method for every encoded value where we need country-dependent behaviour. Do you have an idea how we could achieve this?

Reply:
IMHO only a very limited subset of the encoded values is subject to spatial rules. For example, it wouldn't make sense to override or provide a fallback value for the road surface of a specific area . All the other encoded values fall into the same category as they are basically physical limits(e.g `max_height`) or are to specific to be covered by a countable amount of rules(e.g `hazmat`).

The next step is going to be the introduction of spatial rule files. Maybe in the style of the custom weighting:

```yml
id: DEU
priority: 100

max_speed:
  motor_vehicle:
    road_class:
      motorway: "unlimited"
      trunk: "unlimited"
      primary: 100
      secondary: 100
      tertiary: 100
      residential: 100
      unclassified: 100
      living_street: 4
      "*": "unknown"

road_access:
  motor_vehicle:
    road_class:
      track: "destination"
      path: "no"
      bridleway: "no"
      cycleway: "no"
      footway: "no"
      pedestrian: "no"
      "*": "yes"

toll:
  motor_vehicle:
    road_class:
      motorway: "hgv"
      trunk: "hgv"
      primary: "hgv"
      "*": "no"

area:
  { "type": "Feature", "properties": {},
     "geometry": { 
       "type": "Polygon", 
       "coordinates": [[
       ...
       ]]
     }
  }
```

Reply:
> For example, it wouldn't make sense to override or provide a fallback value for the road surface of a specific area 

Sure. Still it would make sense for use to have a method e.g. for average speed value guessing (car.average_speed) and probably a few others. Probably a method like

```
Object getValue(String valueType, ReaderWay way)
```

would make it more flexible? Or we add methods until it gets more clear on how to solve this.

> or are to specific to be covered by a countable amount of rules(e.g hazmat).

what do you mean here? 

> The next step is going to be the introduction of spatial rule files. Maybe in the style of the custom weighting:

I would keep stuff for now inside Java as the rules could get complex. Later we can surely introduce a file-based approach when there are dozens of Java classes of this kind and we start to see similarities but also exceptions.

btw: see also my latest branch regarding increasing flexibility for CustomWeighting: https://github.com/graphhopper/graphhopper/tree/janino_scripting

Reply:
> Sure. Still it would make sense for use to have a method e.g. for average speed value guessing (car.average_speed) and probably a few others. Probably a method like
> 
> ```
> Object getValue(String valueType, ReaderWay way)
> ```
> 
> would make it more flexible? Or we add methods until it gets more clear on how to solve this.

I'll see if i can simplify the interface a bit :wink: 

> 
> > or are to specific to be covered by a countable amount of rules(e.g hazmat).
> 
> what do you mean here?

One could abuse the spatial rules to add hazmat information which are not part of the OSM material. But that would require a huge amount of tiny rules and defeats the intended purpose of this feature.

> I would keep stuff for now inside Java as the rules could get complex. Later we can surely introduce a file-based approach when there are dozens of Java classes of this kind and we start to see similarities but also exceptions.

Judging from my previous attempt to model the spatial rules for all European countries(#1799), we can safely aim for a simple approach which can be expressed with rule files. Being able to provide arbitrary rules and make them discoverable via PathDetails would be really nice. This is currently impossible without creating a fork and extending the Country enum.

> btw: see also my latest branch regarding increasing flexibility for CustomWeighting: https://github.com/graphhopper/graphhopper/tree/janino_scripting

Looks interesting

Reply:
My biggest gripe with this is typesafety as we would need to pass in the current value as an untyped Object.